src/Common/src/System/Text/StringOrCharArray.cs

```diff
@@ -71,12 +71,14 @@ public unsafe bool Equals(StringOrCharArray other)
             if (this.String != null)
             {
                 // String vs String
+                Debug.Assert(String != null && other.String != null);
                 if (other.String != null)
```

Comment:
Not sure why we are asserting here. `String != null` is checked above and other.String != null is checked below. Compiler complaining below?

Reply:
for StringOrCharArray, either String or CharArray should never be null. 
Therefore, the Equals logic seems to have gone segmented to 4 code blocks for which I added Debug.Assert for. 
I simplified the Debug.Asserts in the next commit to make the compiler happy